tests/dragonfly/utility.py

```diff
@@ -15,68 +20,413 @@ def grouper(n, iterable):
 
+        await asyncio.sleep(0.01)
+        its += 1
 
-def batch_fill_data(client: redis.Redis, gen):
-    for group in grouper(BATCH_SIZE, gen):
-        client.mset({k: v for k, v, in group})
 
+class SizeChange(Enum):
+    SHRINK = 0
+    NO_CHANGE = 1
+    GROW = 2
 
-async def batch_fill_data_async(client: aioredis.Redis, gen):
-    for group in grouper(BATCH_SIZE, gen):
-        await client.mset({k: v for k, v in group})
 
+class ValueType(Enum):
+    STRING = 0
+    LIST = 1
+    SET = 2
+    HSET = 3
+    ZSET = 4
 
-def as_str_val(v) -> str:
-    if isinstance(v, str):
-        return v
-    elif isinstance(v, bytes):
-        return v.decode()
-    else:
-        return str(v)
 
+class CommandGenerator:
+    """Class for generating complex command sequences"""
 
-def batch_check_data(client: redis.Redis, gen):
-    for group in grouper(BATCH_SIZE, gen):
-        vals = [as_str_val(v) for v in client.mget(k for k, _ in group)]
-        gvals = [v for _, v in group]
-        assert vals == gvals
+    def __init__(self, target_keys, val_size, batch_size, max_multikey):
+        self.key_cnt_target = target_keys
+        self.val_size = val_size
+        self.batch_size = min(batch_size, target_keys)
+        self.max_multikey = max_multikey
 
-async def batch_check_data_async(client: aioredis.Redis, gen):
-    for group in grouper(BATCH_SIZE, gen):
-        vals = [as_str_val(v) for v in await client.mget(k for k, _ in group)]
-        gvals = [v for _, v in group]
-        assert vals == gvals
+        # Key management
+        self.key_sets = [set() for _ in ValueType]
+        self.key_cursor = 0
+        self.key_cnt = 0
 
-def wait_available(client: redis.Redis):
-    its = 0
-    while True:
-        try:
-            client.get('key')
-            print("wait_available iterations:", its)
-            return
-        except redis.ResponseError as e:
-            assert "Can not execute during LOADING" in str(e)
+        # Grow factors
+        self.diff_speed = 5
+        self.base_diff_prob = 0.2
+        self.min_diff_prob = 0.1
 
-        time.sleep(0.01)
-        its += 1
+    def keys(self):
+        return itertools.chain(*self.key_sets)
 
+    def keys_and_types(self):
+        return ((k, t) for t in list(ValueType) for k in self.set_for_type(t))
 
-async def wait_available_async(client: aioredis.Redis):
-    its = 0
-    while True:
-        try:
-            await client.get('key')
-            print("wait_available iterations:", its)
-            return
-        except aioredis.ResponseError as e:
-            assert "Can not execute during LOADING" in str(e)
+    def set_for_type(self, t: ValueType):
+        return self.key_sets[t.value]
 
-        await asyncio.sleep(0.01)
-        its += 1
+    def add_key(self, t: ValueType):
+        k, self.key_cursor = self.key_cursor, self.key_cursor + 1
+        self.set_for_type(t).add(k)
+        return k
+
+    def randomize_type(self):
```

Comment:
you don't need self here. I think its better to make it a free function, or as part of ValueType class